headers/storage/lock.h

```diff
@@ -0,0 +1,50 @@
+//
+// Created by Sameer Raj on 14/07/24.
+//
+
+#ifndef LOCK_H
+#define LOCK_H
+#include <condition_variable>
+
+
+struct ReadWriteLock {
+    std::mutex mtx;
+    std::atomic_bool isWriterWaiting{false};
+    int readerCount{0};
+    std::atomic_bool isWriteLocked{false};
+    std::condition_variable readCv;
+    std::condition_variable writeCv;
+};
+
+namespace lock {
+    inline void getReadLock(ReadWriteLock& lock) {
+        std::unique_lock lck(lock.mtx);
+        lock.readCv.wait(lck, [&lock] { return !lock.isWriteLocked && !lock.isWriterWaiting; });
+        ++lock.readerCount;
+    }
+
+    inline void releaseReadLock(ReadWriteLock& lock) {
+        std::unique_lock lck(lock.mtx);
+        --lock.readerCount;
+        if(lock.isWriterWaiting) {
+            lock.writeCv.notify_one();
+        } else {
+            lock.readCv.notify_all();
+        }
```

Comment:
Why do we have a condition variable on read lock ? Isn't it a shared lock ?  If there is someone who has already acquired a read lock, all writes wait on the writeCv variable, but who waits on readCv variable?

Reply:
Both readers and writers will have separate queues.
The readers will wait on the readers' queue, and similarly, the writers will wait on writers' queue, and thus these condition variable.

When the write lock is taken, the readers do need to "wait" or more appropriately sleep, the condition of which is: either write lock is taken, or some writer is in the queue (so to now starve the writers, though this may starve the readers).